modules/core/src/main/java/org/apache/ignite/internal/processors/datastructures/GridCacheLockImpl2Fair.java

```diff
@@ -0,0 +1,747 @@
+
+        assert name != null;
+        assert key != null;
+        assert lockView != null;
+
+        this.name = name;
+        this.ctx = lockView.context();
+
+        sync = new Sync(ctx.localNodeId(), key, lockView, ctx, ctx.logger(getClass()));
+    }
+
+    /** {@inheritDoc} */
+    @Override public void close() {
+        // No-op.
+    }
+
+    /** {@inheritDoc} */
+    @Override public String name() {
+        return name;
+    }
+
+    /** {@inheritDoc} */
+    @Override public void lock() throws IgniteException {
+        ctx.kernalContext().gateway().readLock();
+
+        try {
+            sync.acquire();
+        }
+        finally {
+            ctx.kernalContext().gateway().readUnlock();
+        }
+    }
+
+    /** {@inheritDoc} */
+    @Override public void lockInterruptibly() throws IgniteException {
+        ctx.kernalContext().gateway().readLock();
+
+        try {
+            sync.acquireInterruptibly();
+        }
+        catch (InterruptedException e) {
+            throw new IgniteInterruptedException(e);
+        }
+        finally {
+            ctx.kernalContext().gateway().readUnlock();
+        }
+    }
+
+    /** {@inheritDoc} */
+    @Override public boolean tryLock() throws IgniteException {
+        ctx.kernalContext().gateway().readLock();
+
+        try {
+            return sync.tryAcquire();
+        }
+        catch (IgniteCheckedException e) {
+            throw U.convertException(e);
+        }
+        finally {
+            ctx.kernalContext().gateway().readUnlock();
+        }
+    }
+
+    /** {@inheritDoc} */
+    @Override public boolean tryLock(long timeout, TimeUnit unit) throws IgniteException {
+        assert unit != null;
+
+        ctx.kernalContext().gateway().readLock();
+
+        try {
+            return sync.tryAcquire(timeout, unit);
+        }
+        catch (InterruptedException e) {
+            throw new IgniteInterruptedException(e);
+        }
+        finally {
+            ctx.kernalContext().gateway().readUnlock();
+        }
+    }
+
+    /** {@inheritDoc} */
+    @Override public void unlock() throws IgniteInterruptedException {
+        ctx.kernalContext().gateway().readLock();
+
+        try {
+            sync.release();
+        }
+        finally {
+            ctx.kernalContext().gateway().readUnlock();
+        }
+    }
+
+    /** {@inheritDoc} */
+    @Override public int getHoldCount() throws IgniteException {
+        return sync.reentrantCount.get();
+    }
+
+    /** {@inheritDoc} */
+    @Override public boolean isHeldByCurrentThread() throws IgniteException {
+        return sync.reentrantCount.get() > 0;
+    }
+
+    /** {@inheritDoc} */
+    @Override public boolean isLocked() throws IgniteException {
+        try {
+            if (sync.reentrantCount.get() > 0)
+                return true;
+
+            if (sync.isGloballyLocked)
+                return true;
```

Comment:
.... This can be replaced with if( hasQueuedThreads() ) return true

Reply:
fixed